### ISSUE TITLE
TeamCity service message should have assembly name as a part of test name.

### ISSUE THREAD

**Issue by NikolayPianikov** [nunit/nunit#669]
#671
For example ##teamcity[testStarted name='NUnit.Engine.Tests.TestEngineResultTests.CanCreateFromXmlString' captureStandardOutput='true' flowId='0-1111']
##teamcity[testFinished name='NUnit.Engine.Tests.TestEngineResultTests.CanCreateFromXmlString' duration='0.001' flowId='0-1111']
should be:
##teamcity[testStarted name='nunit.core.engine.tests.dll: NUnit.Engine.Tests.TestEngineResultTests.CanCreateFromXmlString' captureStandardOutput='true' flowId='0-1111']
##teamcity[testFinished name='nunit.core.engine.tests.dll: NUnit.Engine.Tests.TestEngineResultTests.CanCreateFromXmlString' duration='0.001' flowId='0-1111']

It is needed when tests have the same name and another original assembly

**Comment by CharliePoole** [comment]
I feel like we are thrashing on this issue. At @chris-smith-zocdoc 's request, we created the flowId, which is unique across all assemblies, to resolve the same problem. If it's not working, we should fix it of course but maybe we need to have a better understanding of the purpose of each of the parameters before making any more changes.

We probably need to have our teamcity build working before doing this as well.

**Comment by chris-smith-zocdoc** [comment]
I made some progress with the [TC build](https://teamcity.jetbrains.com/project.html?projectId=NUnit_NUnit3_BuildAndTest&tab=projectOverview&branch_NUnit_NUnit3_BuildAndTest=__all_branches__) but did find one new issue https://github.com/nunit/nunit/issues/670  

This particular issue 
> when tests have the same name and another original assembly

was actually brought up in https://github.com/nunit/nunit/pull/590#issuecomment-94918882 but haven't been motivated to resolve it. I feel that @NikolayPianikov 's suggestion is the correct solution though.

**Comment by NikolayPianikov** [comment]
See pull request #671
The "flowId" is just "id of thread" and should be unique if tests are processed in parallel. Thus the "flowId" doesn't help when user has several different assemblies which contain some test methods with the same names (including class's name and namespace). User has several results for his tests and have no ability to collate specific result and corresponding test.
I tried to support backward compatibility. All previous nunit.framework versions should work with new NUnit 3 console. I added attribute "assembly" to xml report. This assembly is used to specify additional prefix of test name in the TeamCity service message.

**Comment by chris-smith-zocdoc** [comment]
> I would like to add "start suite" and "stop suite" service messages to handle cases when the names of assemblies are the same (it is actual for multi platform build for example). Are there any objections?

The "start suite" and "stop suite" messages were removed in https://github.com/nunit/nunit/pull/590 because they interfered with the parallel test execution. Is it possible to use them and run tests in parallel? Most of the discussion around this happened in the google group https://groups.google.com/forum/#!topic/nunit-discuss/X2-LDrsxXXc

**Comment by NikolayPianikov** [comment]
As I understood from the discussion we can start within some suite using different TestWorkers and we have now way to assign correct flowId for test suites. We could use following TC service message ##teamcity[flow Started flowId='<new flow id>'] and ##teamcity[flow Started flowId='<new flow id>' parent='<parent flow id>'] to show tests' results correctly in the TeamCity. These messages allow us to specify the tree of test items (test, test suite) properly. I can do it in the scope of another task may be.

**Comment by NikolayPianikov** [comment]
I've checked ##teamcity[flowStarted ...] and it works fine.
But I think this change is not important now - TC uses his own grouping on the client.
And what do you think about "TeamCity service message should have assembly name as a part of test name"?

**Comment by chris-smith-zocdoc** [comment]
> TeamCity service message should have assembly name as a part of test name

Yes I think this is a good change. It allows users to filter by assembly which is nice.

**Comment by CharliePoole** [comment]
Maybe we need a page in one of our wikis about exactly what messages we send (or want to send) for each teamcity event. If someone wants to volunteer to do this we can give you access to create it.

Also, I am beginning to think that our flowid goes too far. We created a flowid that is actually unique for every test and suite. This is probably not serving the purpose for which flowid was intended. Maybe we should try to actually provide what TeamCity asks for: one flowid per thread of execution.

**Comment by NikolayPianikov** [comment]
I can fill this page, but it would be nice to make review after

TC has the bug, related to our situation - unique flowId per test, but I am going to fix it in the TC. FlowId per test is acceptable solution from my point of view. Solution "one flower per thread of execution" - will be better (but it is not so critical), we could schedule this task and I could make.

**Comment by CharliePoole** [comment]
I created a stub page and gave you access to edit the wiki...

https://github.com/nunit/dev/wiki/TeamCity-Service-Messages

I put it under Specifications in the development wiki for now. When we implement, it can become a Technical Note in the nunit wiki.

The unique id per test was intended as a workaround, which we hoped would solve all problems of duplication. That isn't so because TC only uses the flowid for a specific purpose and not for actually identifying unique tests. We could consider using this unique id as a prefix or suffix to every test name, rather than the assembly name. It's a question of what looks best in the reports, I suppose.

Of course, users don't need to use this feature if they prefer not to. They could instead use the report that is generated from the NUnit XML output.

**Comment by CharliePoole** [comment]
In writing up the service messages and how we use them, I suggest you use links to the TeamCity documentation for each one.

**Comment by NikolayPianikov** [comment]
OK, thanks @CharliePoole, I will look

**Comment by NikolayPianikov** [comment]
See PR #684

**Comment by CharliePoole** [comment]
Although @NikolayPianikov has already done two PRs on it, this has never been accepted as an issue before. @NikolayPianikov it's really desirable to get us to agree that something is a problem before trying to solve it. Of course, I understand there is a problem for you here and that's why you're pushing on it, so I'm accepting it for the Beta we are about to release. 

I can't be sure it will actually get into the beta, but I'll follow up on @rprouse review of PR #882 and we can go from there. I'll save my comments for the PR.

**Comment by CharliePoole** [comment]
Some background questions for @NikolayPianikov ...

My understanding is that TeamCity actually uses the output XML from NUnit 2.x. Is that right?

My further understanding is that you are planning to no longer do that for NUnit 3.0 but instead to use these special messages that must be embedded in NUnit. Is that Right?

My conclusion from the above is that JB is downgrading it's level of support for NUnit and requiring NUnit to instead include support for TeamCity. This doesn't seem like a GoodThing to me and it is the reason why I have been hesitant to take some of your changes before. 

Is my view of what JB is doing accurate, according to you? I realize you may feel you're just doing the programming and not want to speak publicly for JetBrains, but if necessary, you can pass these questions on to somebody else.

My hope is that those folks who want to use NUnit under TeamCity will have the same level of support they have had in the past but that NUnit will not get filled with TeamCity-related code in order to do it.

These general questions are separate from the specific PR review, which I will do on #882.

**Comment by NikolayPianikov** [comment]
- TeamCity uses the nunit-launcher to run tests under the NUnit 2. Tha NUnit-launcher uses listeners (Event Listener) to subscribe on events from NUnit. It sends xml messages to notify the TeamCity about progress in the real time. It is obsolete way. But this approach are used to run NUnit tests under the MSBuild too. The TeamCity does not rely on a xml report from the NUnit, because it is not real time.

- You are right. We would like to use TeamCity service messages everywhere. The nunit-launcher is complex solution and it takes a lot of extra responsibilities, so the majority (70%) issues and new requirements can be solved, avoiding using it, and run the test directly via NUnit-console. From our point of view users should understand what is happening and they should have ability to repeat cases just copying a command line string from a build log, it is not possible for the JB NUnit-launcher, it is working like a black box. Also we should keep backword compatibility concerning the naming of tests' results, for example it is important when user will start using NUnit 2 instead NUnit 3, and the main objective of this PR to do it.

- We would like to make the running of tests under the NUnit more transparent. We have another way, we could implement EventListner to do the same things. It is not so hard, but it is still more difficult than the current method. But the main challenge is to do it for the MSBuild and other types of build step. Current way is simplest when user has good integration from the box and it does not depend on the type of build step.

- I can connect my other colleagues to discuss the issue.

- I can assure you that the level of support will be improved. The TeamCity is popular with .net developers, most of which use NUnit to run unit/integration/auto tests. It's a big part of our users.

The NUnit integration it is one of my zones of responsibility in the JB. Also I like the NUnit and I am using it everywhere, except the Java's world :)

**Comment by CharliePoole** [comment]
Thanks for the extended answer. I have guests about to arrive, so this will all have to lay fallow for a few days. However, I'll definitely get back to you and I want to get some form of this PR into the Beta by the end of the week.

**Comment by CharliePoole** [comment]
@NikolayPianikov This is a general comment on your last note...

I guess I misunderstood. My experience with Teamcity was from clients who used it for daily test runs rather than CI. They didn't care about immediate feedback during the run and just used something (an addon maybe?) to transform the NUnit result file to HTML that was displayed. I can see that folks using TC for CI purposes might want to get feedback as the test is running.

Some specifics: 

You write "we should keep backword compatibility concerning the naming of tests' results." This seems like a tall order! NUnit 3.0 is **not** backward compatible with NUnit 2.x. All our users should know this by now, since we have been saying it for a few years. In particulary, there are results that are reported differently, by design, because  they are viewed as improvements. IMO, it's a mistake to try to hide these differences from users, since our documentation is emphasizing them in many cases.

Also, I imagine some users will want to run their NUnit 2.x tests under NUnit 2.6.4, in spite of the fact that NUnit 3.0 is able to run them. I guess you will need to support them.

You could implement an event listener, as you say, if NUnit 3.0 provided an event listener addin, but it doesn't. The event listening in the new framework is purely external. The only events that go outside the framework are xml messages. It is those messages that are currently used by the team city support. For ease of future maintenance, I think anything you do should really be based on those messages, not on anything internal to the framework.

***** 

I'm glad that support will be improved! We will try to get this change in for the beta-5 release so that you can see how it works before the final release. I'll make more detailed comments on the PR.

**Comment by CharliePoole** [comment]
Closing this after merging. We'll do a new issue for the NUnit 2.x tests and any remaining unresolved questions from this issue.
